### PR TITLE
Send explicit {note what=read} upon message receipt.

### PR DESCRIPTION
**Pull request by aforge**

### CODE REVIEW

File: Tinodios/MessageInteractor.swift [hunk 1]
```diff
@@ -392,6 +392,11 @@ class MessageInteractor: DefaultComTopic.Listener, MessageBusinessLogic, Message
     }
     override func onData(data: MsgServerData?) {
         self.loadMessages()
+        if let from = data?.from, let seq = data?.seq, !Cache.getTinode().isMe(uid: from) {
+            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(1)) {
+                self.sendReadNotification(explicitSeq: seq)
```
Comment (or-else) [inline Tinodios/MessageInteractor.swift]: I suspect the events placed into the main dispatch queue survive lifecycle events, like VC being removed. That means the read notification is going to be sent even when the VC is closed immediately. Or it's going to fire even if topic changes. 1 second is a short window, it's unlikely to happen, but still.

File: Tinodios/MessageInteractor.swift [hunk 2]
```diff
@@ -392,6 +392,11 @@ class MessageInteractor: DefaultComTopic.Listener, MessageBusinessLogic, Message
...
+            }
```
Comment (or-else) [inline Tinodios/MessageInteractor.swift]: Another problem with this approach (and it's also a problem on Android) is that if the user receives a large batch of messages, each one of them is going to be acknowledged separately even if only one ack is needed.

Comment (or-else) [reply Tinodios/MessageInteractor.swift]: I fixed it like this: https://github.com/tinode/tindroid/commit/c5fe504c4b0ecde5e4c3061be331f35726e3edb7

Comment (aforge) [reply Tinodios/MessageInteractor.swift]: Modified iOS correspondingly.

File: Tinodios/MessageInteractor.swift (hunk 1 above)
Comment (aforge) [reply Tinodios/MessageInteractor.swift]: Make reference to `self` weak.

File: Tinodios/MessageViewController.swift [hunk 3]
```diff
@@ -336,33 +334,14 @@ class MessageViewController: UIViewController {
...
-                    self.processNotifications()
-                })
-        } else {
-            // Fallback on earlier versions
-            self.noteTimer = Timer.scheduledTimer(
-                timeInterval: 1,
-                target: self,
-                selector: #selector(self.processNotifications), userInfo: nil, repeats: true)
-        }
+        self.interactor?.sendReadNotification(explicitSeq: nil, when: .now())
```
Comment (or-else) [inline Tinodios/MessageViewController.swift]: think it should be `when: .now() + .seconds(1)`

File: Tinodios/MessageInteractor.swift [hunk 4]
```diff
@@ -193,8 +196,18 @@ class MessageInteractor: DefaultComTopic.Listener, MessageBusinessLogic, Message
         }
         return true
     }
-    func sendReadNotification() {
-        topic?.noteRead()
+    func sendReadNotification(explicitSeq: Int? = nil, when deadline: DispatchTime = .now()) {
```
Comment (or-else) [inline Tinodios/MessageInteractor.swift]: It looks like `sendReadNotification` is never called in a way which makes use of ` = .now()`.

Comment (aforge) [reply Tinodios/MessageInteractor.swift]: Removed.

File: Tinodios/MessageViewController.swift (hunk 3 above)
Comment (aforge) [reply Tinodios/MessageViewController.swift]: Done.

**Pull request merged**